File: cpp/108.cpp
#include <vector>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  TreeNode *insert(vector<int> &arr, int l, int r) {
    if (l > r)
      return nullptr;
    int mid = l + (r - l) / 2;
    TreeNode *cur = new TreeNode(arr[mid]);
    cur->left = insert(arr, l, mid - 1);
    cur->right = insert(arr, mid + 1, r);
    return cur;
  }
  TreeNode *sortedArrayToBST(vector<int> &nums) {
    return insert(nums, 0, nums.size() - 1);
  }
};

File: go/108.go
package main

// Definition for a binary tree node.
// type TreeNode struct {
// 	Val   int
// 	Left  *TreeNode
// 	Right *TreeNode
// }

func insert(nums []int, l int, r int) *TreeNode {
	if l > r {
		return nil
	}
	var mid int = l + (r-l)/2
	cur := TreeNode{
		Val:   nums[mid],
		Left:  nil,
		Right: nil,
	}
	cur.Left = insert(nums, l, mid-1)
	cur.Right = insert(nums, mid+1, r)
	return &cur
}

func sortedArrayToBST(nums []int) *TreeNode {
	return insert(nums, 0, len(nums)-1)
}
